### ISSUE TITLE
How should branch filtering work in combination with reltables?

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2147]
Let's say I have a main DITA Map with content like:

``` xml
<topichead navtitle="Windows Installer" keyscope="windows-installation">
  <ditavalref href="topics/getting_started/install_hwu_windows.ditaval">
    <ditavalmeta>
      <dvrResourceSuffix>-hwuwindows</dvrResourceSuffix>
    </ditavalmeta>
  </ditavalref>
  <topicref href="topics/getting_started/install_select_parallel.dita" keys="install_paralel"/>
  <topicref href="topics/getting_started/install_specify.dita" keys="install_specific"/>
  <mapref href="container_reltable.ditamap" processing-role="resource-only" type="map"/>
</topichead>
```

The "container_reltable.ditamap" contains a reltable which links two of the topics together:

``` xml
<reltable>
  <relrow>
    <relcell linking="sourceonly">
      <topicref
        href="topics/getting_started/install_select_parallel.dita"/>
    </relcell>
    <relcell linking="targetonly">
      <topicref href="topics/getting_started/install_specify.dita">
        <topicmeta>
          <linktext>Run on cluster</linktext>
        </topicmeta>
      </topicref>
    </relcell>
  </relrow>
</reltable>
```

But because of the branch filtering the two topics get renamed so there will be no related links between them in the published output.
How could we make this work? Is this a problem with the specification not covering this aspect or with the DITA OT branch filtering implementation?

I also tried in the reltable to use keyrefs to the keys defined for the topics (having a keyscope defined on the entire "Windows Installer" chapter. But this does not seem to work, the reltables seemed to be globally processed.

Link to [a sample DITA project](https://www.oxygenxml.com/forum/files/branchFilterWithReltable.zip).

**Comment by drmacro** [comment]
I would say that related links that use direct URI references are basically undefined--because it is impossible to know the use context intended there is no way for a processor to do the right thing *because it is impossible to know what the right thing is* in that case.

That is--before DITA 1.2 relationship tables were broken for this reason (that is, they never worked and never could work with direct URI references to topics).

Thus in this case whatever the OT does is equally correct and incorrect.

Topicrefs from relationship tables should only use key references and those references *must* be to normal-role topicrefs, otherwise the same lack of use context condition holds. The only improvement with references to resource-only keys is that those keys can at least be in a scope so the system has the ability to generate correct scope-augmented key references.

So it's not that this case is not addressed by the DITA spec, it's that it *can't be addressed* because there is no single behavior that would be correct. That is, it's a "don't do that" condition.

That said, the OT could apply a default behavior of rewriting URI references within a filtered branch to reflect any scope-defined resource augmentation. That would probably produce the intended effect most of the time (because people making direct URI references have already come to depend on the HTML generation behavior which results in some links working "by accident" but in the way that authors expect (or don't realize is only working by accident).

**Comment by raducoravu** [comment]
I agree, unfortunately using keyrefs (which is the most elegant and non-ambiguous approach) in the reltable still does not work.

**Comment by shaneataylor** [comment]
Adding my vote. For me (and I suspect others) branch processing of reltables would be very useful. As an example, I use branch processing to generate branches of help topics for several different LMSes. Most of the content in the branch is common among all the LMSes (thus, branch processing is useful). Yet with the current limitation, I need to maintain separate reltables for each of the LMSes, with largely duplicated content.

**Comment by drmacro** [comment]
When you say "branch processing of reltables" what do you mean precisely?

That is, do you mean having ditavalrefs within topicrefs within reltable cells or you do you mean having references to the base scopes for topicrefs within filtered branches that also define keyscope augmentations?

Consider this map:

<map>
  <topicref keyscope="scope01">
    <ditavalref ...>
      {markup to make the effective scope "scope01-brA" and add "brA" to resource URIs}
   </ditavalref>
   <ditavalref ...>
     {markup to make the effective scope "scope01-brB" and add "brB" to resource URIs}
   </ditavalref>
   <topicref keys="somekey" href="topic-01.dita"/>
</topicef>
<reltable>
  <relrow>
    <relcell>
      <topicref keyref="scope01.somekey"/>
    </relcell>
  </relrow>
  ...
</reltable>
</map>

The reference as authored is clear and resolvable (key "somekey" exists in scope "scope01").

However, after branch filtering is applied, there will be no "scope01", only "scope01-brA" and scope01-brB". (So one could argue that the key reference is actually incorrect as authored but I don't think we ever considered this case when thinking through key scopes and branch filtering.)

What should happen to the keyref in the reltable?

Seems like a reasonable solution would be to duplicate the topicref for each filtered branch.

Alternatively the author could author the reltable topicrefs to reflect the scopes as filtered:

<relcell>
   <topicref keyref="scope-brA.somekey"/>
</relcell>
<relcell>
   <topicref keyref="scope-brB.somekey"/>
</relcell>

Here reflecting the author's desire to relate topics in one branch to topics in another branch.

The key references are perfectly resolvable given the map as authored because the ditavalref markup provides all the information necessary to compute the effective key scopes, so there should be no reason authors can't reliably create this kind of key reference.

**Comment by shaneataylor** [comment]
I don't see your map, but I can at least partly clarify: what would be useful to me as a writer would be the ability for a relationship table to be processed as part of the branch contents, that is:
```
<topicref...>
  <ditavalref...>
  <ditavalref...>
  <mapref...>
</topicref>
```
where the ditamap referenced by the mapref includes one or more relationship tables.

I hadn't considered ditavalrefs within topicrefs within reltable cells; I don't think that would be useful to me, as the principal value of branch processing is to reuse a significant set of content with variations that can be specified via filtering and scoped keys.

I don't see a situation where defining branches within a relationship table would solve a problem that couldn't be addressed by referencing a specific scope or by including a relationship table within a branch. So, to me that use case adds complexity without value.

That said, because the spec does not prohibit ditavalrefs within topicrefs within reltable cells, if we make an implementation choice not to support this, we should document it and recommend the supported implementation.

**Comment by drmacro** [comment]
By "processed as part of the branch contents" I assume you mean that:

- Any filtering applied to the branch is also applied to the relationship table (e.g., if there are two filters I get two effective relationship tables)
- Any key references are resolved in the context of the branch's key scopes, rather than in the context of the root map's key scopes, meaning that you could have unqualified (or partially-qualified key references in the relationship table that would be resolved in the context of the containing scopes).

If that's what you mean that seems like a reasonable expectation and useful behavior, if not the only useful behavior (the other being to treat the relationship table as though it had occurred as a literal child of the root map, but that seems not at all useful).

However, I'm not sure the DITA spec allows it or explicitly disallows it. Here's what I can find that's relevant:

From the <reltable> reference entry:

"Within a map tree, the effective relationship table is the union of all relationship tables in the map."

That doesn't really address the issue of effective key scopes for relationship tables in maps included in scopes.

I can't find any other discussion of relationship tables and key scopes.

I'm not sure we ever considered the case of a filtered branch including reference to a submap that then contained a relationship table (maybe Robert did)? If we had I'm pretty sure we would have realized the need to say something about it.

This looks like a hole in the specification to me in that we never considered the case and therefore never provided any language to address it.

I think the only way it would make practical and useful sense to have reltables referenced from filtered branches would be if the both:

- The branch filtering is applied to the reltable
- Any key references within such a reltable are treated as though they originated in the containing scope.

Note that this is really more about references to maps that contain reltables where the map reference is within a key scope--the use or non-use of branch filtering doesn't add anything here.

So maybe the TC needs to consider this case and offer some guidance on how we think processors should handle the case in the absence of any clear normative language.

I certainly wouldn't object if the OT decided to implement reltable processing in this way. I can't think of any reason why this wouldn't be the correct behavior.

**Comment by shaneataylor** [comment]
Yes, Eliot, that's precisely what I mean. 😄

**Comment by raducoravu** [comment]
Adding an example with using reltables with keyscopes and branch filtering:
[ditavalRefWithLinkingReltable.zip](https://github.com/dita-ot/dita-ot/files/2565149/ditavalRefWithLinkingReltable.zip)

**Comment by raducoravu** [comment]
👍 Another one of our end users expected branch filtering to work in relationship with related links (direct links, not keyrefs).

**Comment by amyers367** [comment]
Hi,

I have a situation similar to what Radu posted above. I have a ditavalref and a reltable that uses direct links. As others have noted, this does not work - but only for webhelp. For eclipse output, this seems to be working fine.

`<map id=" " title=" ">
<title></title> 
<topicmeta></topicmeta>
<ditavalref href="_oxygen/projects/templates/ditaval/file.ditaval"/>
<topicref>...
`

**Comment by jelovirt** [comment]
Fixed by #2393
